src/VisualStudio/Core/Def/Implementation/TaskList/CommentTaskTokenSerializer.cs

```diff
@@ -15,16 +15,16 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.TaskList
     internal class CommentTaskTokenSerializer : IOptionPersister
     {
         private readonly ITaskList _taskList;
-        private readonly IOptionService _optionService;
+        private readonly IGlobalOptionService _globalOptionService;
 
         private string _lastCommentTokenCache = null;
 
         [ImportingConstructor]
         public CommentTaskTokenSerializer(
-            VisualStudioWorkspace workspace,
+            IGlobalOptionService globalOptionService,
             [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
         {
-            _optionService = workspace.Services.GetService<IOptionService>();
+            _globalOptionService = globalOptionService;
```

Comment:
We want to discourage features attached to a specific workspace to avoid changing options via the global options service. What was the motivation behind fetching/updating this option via global option service?

Reply:
These changes move it to use the same pattern as the `LanguageSettingsPersister`. Requesting the IOptionService during workspace construction would create a cycle since this previously depended on the VisualStudioWorkspace. The OptionService is used to determine whether to add the register EditorConfigDocumentOptionsProvider during workspace construction.

Reply:
Would the approach used in https://github.com/dotnet/roslyn/pull/40289/files#diff-9f934ab20d3d1c874cf2caec75cf98f6 resolve the cycle? Basically, this way no one except the Workspace will access the option service avoiding any cycles here.

Reply:
I believe the problem is that the Workspace constuctor is trying to get an Option value from the OptionService, which causes these IOptionPersisters to be constructed. If the VisualStudioWorkspace is what is being constructed, then it can't be imported.

Reply:
> I believe the problem is that the Workspace constuctor is trying to get an Option value from the OptionService

Where is this happening? I am unable to find it: https://github.com/dotnet/roslyn/blob/d25991bb47e7c5dee726d9df9041ed12b4acdf5c/src/Workspaces/Core/Portable/Workspace/Workspace.cs#L62-L77

https://github.com/dotnet/roslyn/blob/d25991bb47e7c5dee726d9df9041ed12b4acdf5c/src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspace.cs#L23-L36

Reply:
It is a change introduced in this PR. https://github.com/dotnet/roslyn/pull/40513/files#diff-dda9c6c0844e41f1f128e7da1a127d64R80-R85

The `.TryCreate` uses an Option to determine whether to create the EditorConfigDocumentOptionsProvider or whether to let the Legacy provider get created.

Reply:
We talked offline about this - it would be ideal with fetching an option can be avoided during workspace constuction, but it seems like that is going to be tricky, so the only feasible option seems to be go the route that Joey has chosen.